Standardise option format across the entire rpicam suite (#3636)
raspberrypi/documentation: documentation/asciidoc/computers/camera/rpicam_options_common.adoc

=== Common options

The following options apply across all the `rpicam-apps` with similar or identical semantics, unless otherwise noted.

To pass one of the following options to an application, prefix the option name with `--`. If the option requires a value, pass the value immediately after the option name, separated by a single space. If the value contains a space, surround the value in quotes.

Some options have shorthand aliases, for example `-h` instead of `--help`. Use these shorthand aliases instead of the full option name to save space and time at the expense of readability.

> ==== `help`

Reviewer: I can see what you're trying to do here, by mentioning only-once about prefixing the options with double-dashes... unfortunately people are lazy and they don't read the _entire_ documentation from top to bottom, they just search / skim to find what they're looking for, and then copy-paste what they find. So some (_maybe_ not all?) people seeing this new version of the documentation would either try to use `rpicam-hello help` (and be confused why it doesn't work) or they would assume that they can _only_ use `rpicam-hello -h` (and not realise that `--help` exists).
IMVHO of course :slightly_smiling_face:

Author: I'll give this a real good thinkin'. While true, folks can always rely on the help output if they get stuck. Maybe I should toss a reminder at the top of each section to make it a little easier.

Reviewer: Also, you seem to have removed a lot of the example command-usages. Never underestimate the value of working examples that users can just copy'n'paste?

Alias: `-h`

Prints the full set of options, along with a brief synopsis of each option. Does not accept a value.

==== `version`

Prints out version strings for `libcamera` and `rpicam-apps`. Does not accept a value.

> For example:

Reviewer: Perhaps "Example output:" here would be more explanatory than "For example:" ? :man_shrugging:

                             4056x3040 [10.00 fps - (0, 0)/4056x3040 crop]
----

Sensor mode identifiers have the following form:

----
S<Bayer order><Bit-depth>_<Optional packing> : <Resolution list>
----

For the IMX219 in the above example:

* all modes have an `RGGB` Bayer ordering
* all modes provide either 8-bit or 10-bit CSI2 packed readout at the listed resolutions

> * crop, specified as `(<x>, <y>)/<Width>×<Height>`, where `(x, y)` is the location of the crop window of size width × height in the sensor array. The units remain native sensor pixels, even if the sensor is used in binning or skipping mode.

Reviewer: I don't think this bullet-point is specific to IMX219, so perhaps it ought to be moved above "For the IMX219..." line?

Author: Excellent point, how did I overlook this?

==== `camera`

Selects the camera to use. Specify an index from the xref:camera_software.adoc#list-cameras[list of available cameras].

==== `config`

Alias: `-c`

> Specify a file containing CLI options and values. Consider a file named `config.txt` that contains the following text, specifying options and values as key-value pairs, one option per line:

Reviewer: Minor suggestion, but perhaps it's worth suggesting a filename _other_ than `config.txt` here, in order to avoid any possible confusion with `/boot/firmware/config.txt`? :wink:

Author: Absolutely.

----
timeout=99000
verbose=
----

> TIP: Omit the leading `--` that you normally pass on the command line. For flags that lack a value, such as `verbose` in the above example, you must include a trailing `=`.

Reviewer: IMHO it's probably important to also retain the "Only long form options are permitted (`t=99000` would not be accepted)." warning (or similar) that the original text had.

Author: Agreed.

| %rg
| Gain applied to the red component of each pixel.

| %bg
| Gain applied to the blue component of each pixel.

| %focus
| Focus metric for the image, where a larger value implies a sharper image.

| %lp
| Current lens position in dioptres (1 / distance in metres).

| %afstate
| Autofocus algorithm state (`idle`, `scanning`, `focused` or `failed`).
|===

image::images/focus.jpg[Image showing focus measure]

==== `width` and `height`

Each accepts a single number defining the dimensions, in pixels, of the captured image.

For `rpicam-still`, `rpicam-jpeg` and `rpicam-vid`, specifies output resolution.

For `rpicam-raw`, specifies raw frame resolution. For cameras with a 2×2 binned readout mode, specifying a resolution equal to or smaller than the binned mode captures 2×2 binned raw frames.

For `rpicam-hello`, has no effect.

Examples:

* `rpicam-vid -o test.h264 --width 1920 --height 1080` captures 1080p video.

* `rpicam-still -r -o test.jpg --width 2028 --height 1520` captures a 2028×1520 resolution JPEG. If used with the HQ camera, uses 2×2 binned mode, so the raw file (`test.dng`) contains a 2028×1520 raw Bayer image.

==== `viewfinder-width` and `viewfinder-height`

> Each accepts a single number defining the dimensions, in pixels, of the image sent to for display in the preview window. Does not effect the preview window, since images are resized to fit. Does not effect captured still image or videos.

Reviewer: "sent to for display" ?

Reviewer: effect -> affect ?

Author: ```suggestion
Each accepts a single number defining the dimensions, in pixels, of the image displayed in the preview window. Does not effect the preview window dimensions, since images are resized to fit. Does not affect captured still images or videos.
```

==== `rawfull`

> Forces the sensor to capture images in full resolution mode regardless of xref:camera_software.adoc#width-and-height[requested output resolution]. Because larger resolutions require more resources, can negatively impact framerate. When used with the HQ camera, each frame can take up to 18MB of space (compared with 5MB in 2×2 binned mode). Does not accept a value.

Reviewer: "can negatively" -> "this can negatively" ?

Bit-depth and packing are optional.
Bit-depth defaults to 12.
Packing defaults to `P` (packed).

For information about the bit-depth, resolution, and packing options available for your sensor, see xref:camera_software.adoc#list-cameras[`list-cameras`].

Examples:

* `4056:3040:12:P` - 4056×3040 resolution, 12 bits per pixel, packed.
* `1632:1224:10` - 1632×1224 resolution, 10 bits per pixel.
* `2592:1944:10:U` - 2592×1944 resolution, 10 bits per pixel, unpacked.
* `3264:2448` - 3264×2448 resolution.

===== Packed format details

The packed format uses less storage for pixel data.

_On Raspberry Pi 4 and earlier devices_, the packed format packs pixels using the MIPI CSI-2 standard. This means:

* 10-bit camera modes pack 4 pixels into 5 bytes. The first 4 bytes contain the 8 most significant bits (MSBs) of each pixel, and the final byte contains the 4 pairs of least significant bits (LSBs).
* 12-bit camera modes pack 2 pixels into 3 bytes. The first 2 bytes contain the 8 most significant bits (MSBs) of each pixel, and the final byte contains the 4 least significant bits (LSBs) of both pixels.

> _On Raspberry Pi 5 and later devices_, the packed format compress pixel values with a visually lossless compression scheme into 8 bits (1 byte) per pixel.

Reviewer: compress -> compresses ?

Author: ```suggestion
_On Raspberry Pi 5 and later devices_, the packed format compresses pixel values with a visually lossless compression scheme into 8 bits (1 byte) per pixel.
```

===== Unpacked format details

The unpacked format provides pixel values that are much easier to manually manipulate, at the expense of using more storage for pixel data.

> On all devices, the unpacked format use 2 bytes per pixel.

Reviewer: use -> uses ?

Author: ```suggestion
On all devices, the unpacked format uses 2 bytes per pixel.
```

==== `viewfinder-mode`

Identical to the `mode` option, but it applies to the data passed to the preview window. For more information, see the xref:camera_software.adoc#mode[`mode` documentation].

==== `lores-width` and `lores-height`

Delivers a second, lower-resolution image stream from the camera, scaled down to the specified dimensions.

Each accepts a single number defining the dimensions, in pixels, of the lower-resolution stream.

Available for preview and video modes. Not available for still captures. If you specify a aspect ratio different from the normal resolution stream, generates non-square pixels.

For `rpicam-vid`, disables extra colour-denoise processing.

Useful for image analysis when combined with xref:camera_software.adoc#post-processing[image post-processing].

==== `hflip`

Flips the image horizontally. Does not accept a value.

==== `vflip`

Flips the image vertically. Does not accept a value.

==== `rotation`

Rotates the image extracted from the sensor. Accepts only the values 0 or 180.

==== `roi`

Crops the image extracted from the full field of the sensor. Accepts a proportion of the available field of view in the following format: `<x>,<y>,<w>,h>`

> These value define the following proportions:

Reviewer: value -> values ?

Author: ```suggestion
These values define the following proportions:
```

> * `<x>`: X coordinates to skip before extracting an image

Reviewer: IMHO these descriptions need to make clearer that they're a fraction of the complete image width (in the range 0 - 1), and not a count of pixels.

* `<y>`: Y coordinates to skip before extracting an image
* `<w>`: image width to extract
* `<h>`: image height to extract

Defaults to `0,0,1,1` (starts at the first X coordinate and the first Y coordinate, uses 100% of the image width, uses 100% of the image height).

Examples:

* `rpicam-hello --roi 0.25,0.25,0.5,0.5` selects exactly a half of the total number of pixels cropped from the centre of the image (skips the first 25% of X coordinates, skips the first 25% of Y coordinates, uses 50% of the total image width, uses 50% of the total image height).
* `rpicam-hello --roi 0,0,0.25,0.25` selects exactly a quarter of the total number of pixels cropped from the top left of the image (skips the first 0% of X coordinates, skips the first 0% of Y coordinates, uses 25% of the image width, uses 25% of the image height).

==== `hdr`

Default value: `off`

Runs the camera in HDR mode. If passed without a value, assumes `auto`. Accepts one of the following values:

* `off` - Disables HDR.

> * `auto` - Enables HDR on supported devices. Uses the sensor's built-in HDR mode if available. If the sensor lacks a built-in HDR mode, uses on-board HDR mode, if available.

Reviewer: IMHO "If the sensor lacks a built-in HDR mode, uses on-board HDR mode, if available." is a bit hard to understand.

* `single-exp` - Uses on-board HDR mode, if available, even if the sensor has a built-in HDR mode. If on-board HDR mode is not available, disables HDR.

> Raspberry Pi 5 and later devices have an on-board HDR mode.

Reviewer: IMHO the original text of "on-chip" might be slightly clearer here, because "on-board" feels like a much looser term?
"On-board" could mean _somewhere_ on the Pi5's PCB, or _somewhere_ on the Camera's PCB, whereas "on-chip" feels much more obviously something to do with BCM2712 or RP1.

OTOH perhaps that's a level of distinction that non-engineers don't need? :shrug: :confused:

> To check for built-in HDR modes in a sensor, pass this option in addition to xref:camera_software.adoc#list-cameras[`list-cameras`].

Reviewer: Huh, that's weird - the original text suggested using `rpicam-hello --hdr sensor --list-cameras`, but didn't list `sensor` as a valid argument for the `--hdr` option?

Author: I can't find it in the source code, guessing it was removed?

Reviewer: Or it was a typo in the original documentation? :shrug:

* `1.0` applies the default amount of sharpening
* values greater than `1.0` apply extra sharpening

==== `contrast`

Specifies the image contrast. Accepts a numeric value along the following spectrum:

* `0.0` applies minimum contrast
* values greater than `0.0`, but less than `1.0` apply less than the default amount of contrast
* `1.0` applies the default amount of contrast
* values greater than `1.0` apply extra contrast

==== `brightness`

Specifies the image brightness, added as an offset to all pixels in the output image. Accepts a numeric value along the following spectrum:

* `-1.0` applies minimum brightness (black)
* `0.0` applies standard brightness
* `1.0` applies maximum brightness (white)

For many use cases, prefer xref:camera_software.adoc#ev[`ev`].

==== `saturation`

Specifies the image colour saturation. Accepts a numeric value along the following spectrum:

* `0.0` applies minimum saturation (grayscale)
* values greater than `0.0`, but less than `1.0` apply less than the default amount of saturation
* `1.0` applies the default amount of saturation
* values greater than `1.0` apply extra saturation

==== `ev`

> Specifies the https://en.wikipedia.org/wiki/Exposure_value[exposure value (EV)] compensation of the image in stops. Accepts a numeric value that controls target values passed to the Automatic Gain/Exposure Control (AEC/AGC) processing algorithm along the following spectrum:

Reviewer: Perhaps "Gain/Exposure" should be listed in the same order as "AEC/AGC"? :wink:

Sometimes, digital gain can exceed 1.0 even when the analogue gain limit is not exceeded. This can occur in the following situations:

* Either of the colour gains drops below 1.0, which will cause the digital gain to settle to 1.0/min(red_gain,blue_gain). This keeps the total digital gain applied to any colour channel above 1.0 to avoid discolouration artefacts.
* Slight variances during Automatic Gain/Exposure Control (AEC/AGC) changes.

==== `metering`

Default value: `centre`

Sets the metering mode of the Automatic Gain/Exposure Control (AEC/AGC) algorithm. Accepts the following values:

* `centre` - centre weighted metering
* `spot` - spot metering
* `average` - average or whole frame metering
* `custom` - custom metering mode defined in the camera tuning file

For more information on defining a custom metering mode, and adjusting region weights in existing metering modes, see the https://datasheets.raspberrypi.com/camera/raspberry-pi-camera-guide.pdf[Tuning guide for the Raspberry Pi cameras and libcamera].

==== `exposure`

> Sets the exposure profile. Changing the exposure profile should not effect the image exposure. Instead, different modes adjust gain settings to achieve the same net result. Accepts the following values:

Reviewer: effect -> affect ?

* `sport`: short exposure, larger gains
* `normal`: normal exposure, normal gains
* `long`: long exposure, smaller gains

You can edit exposure profiles using tuning files. For more information, see the https://datasheets.raspberrypi.com/camera/raspberry-pi-camera-guide.pdf[Tuning guide for the Raspberry Pi cameras and libcamera].

==== `awb`

> Sets the Auto White Balance (AWB) algorithm. Accepts the following values:

Reviewer: algorithm -> mode ?

* `continuous`: adjusts the lens position automatically as the scene changes

This option is only supported for certain camera modules.

==== `autofocus-range`

Default value: `normal`

Specifies the autofocus range. Accepts the following values:

* `normal`: focuses from reasonably close to infinity
* `macro`: focuses only on close objects, including the closest focal distances supported by the camera
* `full`: focus on the entire range, from the very closest objects to infinity

This option is only supported for certain camera modules.

==== `autofocus-speed`

Default value: `normal`

Specifies the autofocus speed. Accepts the following values:

* `normal`: changes the lens position at normal speed
* `fast`: changes the lens position quickly

This option is only supported for certain camera modules.

==== `autofocus-range`

Specifies the autofocus window within the full field of the sensor. Accepts a proportion of the available field of view in the following format: `<x>,<y>,<w>,h>`

These value define the following proportions:

> * `<x>`: X coordinates to skip before applying autofocus

Reviewer: Same comment applies here as earlier
